packages/react-reconciler/src/ReactFiberHotReloading.js

```diff
@@ -0,0 +1,312 @@
+  if (__DEV__) {
+    if (fibersWithIgnoredDependencies === null) {
+      // Not hot reloading now.
+      return false;
+    }
+    return fibersWithIgnoredDependencies.has(fiber);
+  } else {
+    return false;
+  }
+}
+
+export function scheduleHotUpdate(hotUpdate: HotUpdate): void {
+  if (__DEV__) {
+    // TODO: warn if its identity changes over time?
+    familiesByType = hotUpdate.familiesByType;
+
+    const {root, staleFamilies, updatedFamilies} = hotUpdate;
+    flushPassiveEffects();
+    fibersWithForcedRender = new Set();
+    fibersWithIgnoredDependencies = new Set();
+    try {
+      flushSync(() => {
+        scheduleFibersWithFamiliesRecursively(
+          root.current,
+          updatedFamilies,
+          staleFamilies,
+        );
+      });
+    } finally {
+      fibersWithForcedRender = null;
+      fibersWithIgnoredDependencies = null;
+    }
+    flushPassiveEffects();
+  }
+}
+
+function scheduleFibersWithFamiliesRecursively(
+  fiber: Fiber,
+  updatedFamilies: Set<Family>,
+  staleFamilies: Set<Family>,
+) {
+  if (__DEV__) {
+    const {child, sibling, tag, type} = fiber;
+
+    let candidateType = null;
+    switch (tag) {
+      case FunctionComponent:
+      case SimpleMemoComponent:
+        candidateType = type;
+        break;
+      case ForwardRef:
+        candidateType = type.render;
+        break;
+      default:
+        break;
+    }
+
+    if (fibersWithForcedRender === null) {
+      throw new Error(
+        'Expected fibersWithForcedRender to be set during hot reload.',
+      );
+    }
+    if (fibersWithIgnoredDependencies === null) {
+      throw new Error(
+        'Expected fibersWithIgnoredDependencies to be set during hot reload.',
+      );
+    }
+    if (familiesByType === null) {
+      throw new Error('Expected familiesByType to be set during hot reload.');
+    }
+
+    if (candidateType !== null) {
+      const family = familiesByType.get(candidateType);
+      if (family !== undefined) {
+        if (staleFamilies.has(family)) {
+          let fiberToRemount = fiber;
+          // Force a remount by changing the element type.
+          // If necessary, do this for more than a single fiber on parent path.
+          while (true) {
+            fiberToRemount.elementType = 'DELETED';
+            if (fiberToRemount.alternate !== null) {
+              fiberToRemount.alternate.elementType = 'DELETED';
+            }
+            const parent = fiberToRemount.return;
+            if (parent !== null && parent.tag === MemoComponent) {
+              // Memo components can't reconcile themelves so
+              // delete them too until we find a non-memo parent.
+              // TODO: we might need something similar for Suspense
+              // or its special "fake fragment" node.
+              fiberToRemount = parent;
+            } else {
+              break;
+            }
+          }
+          // Schedule the parent.
+          const parent = fiberToRemount.return;
```

Comment:
Can you explain to me why the parents need to force re-render? Isn't it sufficient that the component whose code changed re-renders without bailing out?

(Sorry I didn't review this part last time; I must have skimmed over it for some reason.)

Reply:
Your gist includes this line

> Currently there are some bailouts that are impossible to override (like on a root).

but that doesn't apply to deep updates. Can you clarify what you mean here?

Reply:
Note there are two different codepaths.

```js
if (staleFamilies.has(family)) {
  fiberToRemount.elementType = 'DELETED';
  // schedule parent
} else if (updatedFamilies.has(family)) {
  // schedule self
}
```

The first code path is about *forcing remount (despite equivalent type)*, whereas the second path is about *forcing re-render*.

Forcing remount is necessary in case you change order of Hooks. We could also later use that to force remount edited classes. To force remount, we mutate `elementType` to `'DELETED'` and *schedule the parent*.

For normal edits, we don't force remount. We just force re-render of same fiber.

Reply:
>but that isn't relevant to deep updates (setState). Can you clarify what you mean here?

This was about forcing remount of a child. (Which needs a re-render of the parent.) Indeed if we follow a different approach for forcing remount, this comment is outdated.